server/sonar-server/src/main/java/org/sonar/server/computation/period/PeriodsRepository.java

```diff
@@ -0,0 +1,187 @@
+ */
+
+package org.sonar.server.computation.period;
+
+import com.google.common.base.Strings;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.apache.commons.lang.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.api.CoreProperties;
+import org.sonar.api.config.Settings;
+import org.sonar.api.resources.Qualifiers;
+import org.sonar.api.utils.DateUtils;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.core.component.ComponentDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.server.computation.batch.BatchReportReader;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.db.DbClient;
+
+/**
+ * Repository of periods used to compute differential measures.
+ * Here are the steps to retrieve these periods :
+ * - Read the 5 period properties ${@link CoreProperties#TIMEMACHINE_PERIOD_PREFIX}
+ * - Try to find the matching snapshots from the properties
+ * - If a snapshot is found, a new period is added to the repository
+ */
+public class PeriodsRepository {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PeriodsRepository.class);
+
+  private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat(DateUtils.DATE_FORMAT);
+
+  private static final int NUMBER_OF_PERIODS = 5;
+
+  private final DbClient dbClient;
+  private final Settings settings;
+  private final TreeRootHolder treeRootHolder;
+  private final PeriodFinder periodFinder;
+  private final BatchReportReader batchReportReader;
+
+  private List<Period> periods = new ArrayList<>();
+
+  public PeriodsRepository(DbClient dbClient, Settings settings, TreeRootHolder treeRootHolder, PeriodFinder periodFinder, BatchReportReader batchReportReader) {
+    this.dbClient = dbClient;
+    this.settings = settings;
+    this.treeRootHolder = treeRootHolder;
+    this.periodFinder = periodFinder;
+    this.batchReportReader = batchReportReader;
+  }
+
+  public List<Period> getPeriods(){
+   if (periods.isEmpty()) {
+      initPeriods();
+    }
+    return periods;
+  }
+
+  private void initPeriods() {
+    DbSession session = dbClient.openSession(false);
+    try {
+      Component project = treeRootHolder.getRoot();
+      ComponentDto projectDto = dbClient.componentDao().selectNullableByKey(session, project.getKey());
+      // No project on first analysis, no period
+      if (projectDto != null) {
+        BatchReport.Component batchProject = batchReportReader.readComponent(project.getRef());
+        PeriodResolver periodResolver = new PeriodResolver(session, projectDto.getId(), batchReportReader.readMetadata().getAnalysisDate(), batchProject.getVersion(),
+          // TODO qualifier will be different for Views
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)